engineio/src/asynchronous/async_socket.rs

```diff
@@ -80,38 +83,74 @@ impl Socket {
         Ok(())
     }
 
-    /// Creates a stream over the incoming packets, uses the streams provided by the
-    /// underlying transport types.
-    pub(crate) fn stream(&self) -> Result<impl Stream<Item = Result<Option<Packet>>> + '_> {
-        if !self.connected.load(Ordering::Acquire) {
-            return Err(Error::IllegalActionBeforeOpen());
+    /// A helper method that distributes
+    pub(super) async fn handle_inconming_packet(&self, packet: Packet) -> Result<()> {
+        // check for the appropriate action or callback
+        self.handle_packet(packet.clone());
+        match packet.packet_id {
+            PacketId::MessageBinary => {
+                self.handle_data(packet.data.clone());
+            }
+            PacketId::Message => {
+                self.handle_data(packet.data.clone());
+            }
+            PacketId::Close => {
+                self.handle_close();
+            }
+            PacketId::Open => {
+                unreachable!("Won't happen as we open the connection beforehand");
```

Comment:
nit: consider throwing InvalidPacket. if there is a bug in the lib here, panicking would be annoying for library consumers (ideally libraries never panic, just error)

Reply:
Yeah good point! ;)